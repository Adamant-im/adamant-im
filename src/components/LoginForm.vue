<template>
  <v-form v-model="validForm" @submit.prevent="submit" ref="form" class="login-form">

    <v-layout>
      <slot>
        <v-text-field
          v-model="passphrase"
          :label="$t('login.password_label')"
          browser-autocomplete="current-password"
          class="text-xs-center"
          color="grey"
          type="password"
        />
      </slot>

      <slot name="append-outer"></slot>
    </v-layout>

    <v-layout row wrap align-center justify-center>
      <slot name="button">
        <v-btn
          :disabled="!validForm || disabledButton"
          class="login-button v-btn--primary"
          @click="submit"
        >
          <v-progress-circular
            v-show="showSpinner"
            indeterminate
            color="primary"
            size="24"
            class="mr-3"
          />
          {{ $t('login.login_button') }}
        </v-btn>
      </slot>
    </v-layout>

    <transition name="slide-fade">
      <v-layout justify-center>
        <slot name="qrcode-renderer"/>
      </v-layout>
    </transition>
  </v-form>
</template>

<script>
export default {
  computed: {
    passphrase: {
      get () {
        return this.value
      },
      set (value) {
        this.$emit('input', value)
      }
    }
  },
  data: () => ({
    validForm: true,
    disabledButton: false,
    showSpinner: false
  }),
  methods: {
    submit () {
      if (!this.$refs.form.validate()) return false

      this.freeze()
      this.login()
    },
    login () {
      const passphrase = this.passphrase.toLowerCase().trim()

      const promise = this.$store.dispatch('login', passphrase)

      promise
        .then(() => {
          this.$emit('login')
        })
        .catch(() => {
          this.$emit('error', 'login.invalid_passphrase')
        })
        .finally(() => {
          this.antiFreeze()
        })

      return promise
    },
    freeze () {
      this.disabledButton = true
      this.showSpinner = true
    },
    antiFreeze () {
      this.disabledButton = false
      this.showSpinner = false
    }
  },
  props: {
    value: {
      type: String,
      default: ''
    }
  }
}
</script>

<style lang="stylus" scoped>
/**
 * Centering input text and label.
 *
 * 1. Override `style` attribute.
 * 2. Align input text to center.
 * 3. Fix label centering after `scaleY` animation, using `transition font-size` instead.
 */
.login-form >>> .v-label { /* [1] */
  width: 100% !important;
  max-width: 100% !important;
  left: 0;
}
.login-form >>> .v-input input {
  text-align: center; /* [2] */
}
.login-form >>> .v-input .v-label--active { /* [3] */
  transition: font .3s ease;
  transform: translateY(-18px);
  -webkit-transform: translateY(-18px);
  font-size: 12px;
}

.login-button
  min-width: 126px
</style>
