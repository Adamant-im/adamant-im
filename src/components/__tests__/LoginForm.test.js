import { shallowMount } from '@vue/test-utils'
import Vue from 'vue'
import Vuex from 'vuex'
import VueI18n from 'vue-i18n'
import Vuetify from 'vuetify'
import LoginForm from '@/components/LoginForm'

import mockupI18n from './__mocks__/plugins/i18n'
import loginMock from './__mocks__/login'

Vue.use(Vuex)
Vue.use(VueI18n)
Vue.use(Vuetify)

// Because Node.js is not supporting Promise.finally.
// In the future, polyfill can be added.
// eslint-disable-next-line
Promise.prototype.finally = Promise.prototype.finally || {
  finally (fn) {
    const onFinally = cb => Promise.resolve(fn()).then(cb)
    return this.then(
      result => onFinally(() => result),
      reason => onFinally(() => { throw reason })
    )
  }
}.finally

/**
 * Mockup store helper
 */
function mockupStore () {
  const state = () => ({
    address: '',
    balance: 0,
    passphrase: ''
  })

  const getters = {
    isLogged: jest.fn()
  }

  const mutations = {
    setAddress: jest.fn(),
    setBalance: jest.fn(),
    setPassphrase: jest.fn()
  }

  const actions = {
    login: loginMock,
    logout: jest.fn()
  }

  const store = new Vuex.Store({
    state,
    getters,
    mutations,
    actions
  })

  return {
    store,
    state,
    getters,
    mutations,
    actions
  }
}

describe('LoginForm.vue', () => {
  let i18n = null
  let store = null

  beforeEach(() => {
    const vuex = mockupStore()

    store = vuex.store

    i18n = mockupI18n()
  })

  it('renders the correct markup', () => {
    const wrapper = shallowMount(LoginForm, {
      i18n,
      store
    })

    expect(wrapper.element).toMatchSnapshot()
  })

  it('should login with passphrase', async () => {
    const wrapper = shallowMount(LoginForm, {
      i18n,
      store
    })

    wrapper.setProps({ value: 'correct passphrase' })

    const promise = wrapper.vm.login()
    await expect(promise).resolves.toEqual(true)

    expect(wrapper.emitted().login).toBeTruthy()
  })

  it('should throw error when login with incorrect passphrase', async () => {
    const wrapper = shallowMount(LoginForm, {
      i18n,
      store
    })

    wrapper.vm.passphrase = 'incorrect passphrase'

    const promise = wrapper.vm.login()

    try {
      await promise
    } catch (e) {
      expect(e.message).toEqual('Incorrect passphrase')
    }
  })

  it('should freeze form when calling freeze()', () => {
    const wrapper = shallowMount(LoginForm, {
      i18n,
      store
    })

    // default values
    expect(wrapper.vm.disabledButton).toBe(false)
    expect(wrapper.vm.showSpinner).toBe(false)

    // freeze
    wrapper.vm.freeze()
    expect(wrapper.vm.disabledButton).toBe(true)
    expect(wrapper.vm.showSpinner).toBe(true)

    // antifreeze
    wrapper.vm.antiFreeze()
    expect(wrapper.vm.disabledButton).toBe(false)
    expect(wrapper.vm.showSpinner).toBe(false)
  })
})
