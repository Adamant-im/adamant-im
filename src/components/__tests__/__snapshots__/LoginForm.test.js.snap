// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LoginForm.vue > renders the correct markup 1`] = `
<v-form
  class="login-form"
  modelvalue="true"
>
  <v-row
    no-gutters=""
  >
    
    <v-text-field
      autocomplete="current-password"
      class="text-center"
      color="#BBDEFB"
      label="Enter your passphrase to login"
      modelvalue=""
      type="password"
      variant="underlined"
    />
    
    
    
  </v-row>
  <v-row
    align="center"
    justify="center"
    no-gutters=""
  >
    
    <v-btn
      class="login-form__button a-btn-primary"
      disabled="false"
    >
      <v-progress-circular
        class="mr-4"
        color="primary"
        indeterminate=""
        size="24"
        style="display: none;"
      />
       Login
    </v-btn>
    
  </v-row>
  <transition-stub
    appear="false"
    css="true"
    name="slide-fade"
    persisted="false"
  >
    <v-row
      justify="center"
      no-gutters=""
    >
      
      
    </v-row>
  </transition-stub>
</v-form>
`;
